enhancements/baremetal/coreos-image-in-release.md

```diff
@@ -0,0 +1,258 @@
+
+Following a similar approach to that currently used with the QCOW2 also
+effectively extends the limitation that we are not updating the provisioning OS
+image to include the deploy image as well (although we will continue to be able
+to update IPA itself).
+
+We wish to avoid these problems by obtaining an up-to-date CoreOS from the
+release payload.
+
+### Goals
+
+* Ensure that no matter which version of OpenShift a cluster was installed
+  with, we are able to deliver updates to IPA and the OS it runs on.
+* Stop maintaining and shipping the non-CoreOS, RHEL-based IPA PXE files.
+* Never break existing clusters, even if they are deployed in disconnected
+  environments.
+* Eliminate the extra reboot required to update CoreOS after initial
+  provisioning.
+
+### Non-Goals
+
+* Automatically switch pre-existing MachineSets to deploy with `coreos-install`
+  instead of via QCOW2 images.
+* Update the CoreOS QCOW2 image in the cluster with each OpenShift release.
+* Build CoreOS images for platforms other than baremetal.
+
+## Proposal
+
+Build a container image containing the latest CoreOS ISO and the
+coreos-installer binary. This container can be used e.g. as an init container
+to make the ISO available where required. The `coreos-installer iso extract
+pxe` command can also be used to produce the kernel, initrd, and rootfs from
+the ISO, for the purposes of PXE booting.
+
+This image could either replace the existing content of the
+`ironic-ipa-downloader` repo, be built from the new
+`image-customization-controller` repo, or be built from a new repo.
+
+### User Stories
+
+As an operator of a disconnected cluster, I want to upgrade my cluster and have it to continue to work for provisioning baremetal machines.
+
+As an operator of an OpenShift cluster, I want to add to my cluster new
+hardware that was not fully supported in RHEL at the time I installed the
+cluster.
+
+As an operator of an OpenShift cluster, I want to ensure that the OS running on
+hosts prior to them being provisioned as part of the cluster is up to date with
+bug and security fixes.
+
+### Implementation Details/Notes/Constraints
+
+We will need to restore a [change to the Machine Config
+Operator](https://github.com/openshift/machine-config-operator/pull/1792) to
+allow working with different versions of Ignition that was [previously
+reverted](https://github.com/openshift/machine-config-operator/pull/2126) but
+should now be viable after [fixes to the
+installer](https://github.com/openshift/installer/pull/4413).
+
+The correct version of CoreOS to use is available from the openshift-installer;
+this could be obtained by using the installer container image as a build image.
+
+Because OpenShift container images are built in an offline environment, it is
+not possible to simply download the image from the public Internet at container
+build time. Instead this will be accomplished by *MAGIC*. Uploading the ISOs to
+the lookaside cache in Brew is one possibility.
```

Comment:
Do we have a plan for resolving exactly how this process will work?

Reply:
We'll need to pull in some folks from the appropriate team (ART?) to tell us the right way to accomplish it.

Reply:
the good news is that there is a mechanism for ART to download and insert binaries like this if that is in fact what we determine to do. we would just need a script written to do it at the time we rebase the source.

the bad news is that it deepens the dependency tree of things that need to align in a release and falls outside of how we currently determine what needs building.

dependency tree: currently we have hyperkube building as an RPM, getting included in an RHCOS build, and that build then getting included in a nightly. this change would mean that for consistency, baremetal would also need to rebuild and be included in a nightly. there are lots of ways for that to go wrong and we would end up with mismatching content. i would say we could validate that before release but i'm actually not sure how we would, would have to think about it. in any case, there would likely be sporadic extended periods where the two were out of sync and we couldn't form a valid release.

detection that the image needs rebuilding: currently we check whether the source has changed, whether the config has changed, or whether RPM contents have changed, which are all things we can easily look up from the image or brew. there's nothing like that for other content. i guess to do this, the process that does the upload could add a label with the RHCOS buildid that's included, and we could then compare that to see if there's a newer build available. i don't look forward to adding that bit of complexity into the scan but it seems solvable.

some other things to consider:
1) how does this work for OKD/CI builds?
2) i assume we need this for all architectures? each arch is a different RHCOS build and that makes it a bit tricky to reuse the same dockerfile for all arches (when the content will need different arch-dependent names).

Reply:
So far baremetal only supports x86_64. But it seems inevitable that _one_ day we'll have to support multiple architectures. Could we have one Dockerfile per arch?

Reply:
I think we should always keep this ISO the same as the one pinned in the installer.

So one idea: have this image derive from the installer.  I think we'd get edge triggering for free then.

Reply:
> So far baremetal only supports x86_64. But it seems inevitable that one day we'll have to support multiple architectures.

Note that the [stream metadata](https://github.com/openshift/installer/blob/master/data/data/rhcos-stream.json) includes multiple architectures already, and I think this should use that as a source of truth.

Reply:
config has baremetal _building_ for all arches so i assumed those were used but it's simpler if not. we should probably limit those builds to x86...

our build system insists that a single multi-arch build reuse the same dockerfile for all (this is useful for ensuring they all have consistent content). there are ways around it, though, involving either complicating the dockerfile a bit, or (more likely) splitting the builds into a build per arch - we already have some like this for ARM64. so I guess this is not a blocker, just another complication.

Reply:
keeping the ISO same as installer seems like a good idea. that would mean it's not changing frequently.

i'm not _quite_ sure how we'd implement that; i'm not sure how we'd determine what to download prior to the build, and we can't download during the build. except maybe from brew... heh.

Reply:
> some other things to consider:
> 
> 1. how does this work for OKD/CI builds?
> 
> 2. i assume we need this for all architectures? each arch is a different RHCOS build and that makes it a bit tricky to reuse the same dockerfile for all arches (when the content will need different arch-dependent names).

1: For OKD, the ostree in machine-os-content is different than the one used in the bootimage (which is FCOS). For the okd-machine-os ostree, the hyperkube and client RPMs are extracted from the artifacts image, and layered onto FCOS: https://github.com/openshift/okd-machine-os/blob/master/Dockerfile.cosa#L9. Since the trees differ, a new node in OKD always has to pivot from FCOS to okd-machine-os before it even has a kubelet do run anything with.  

2: This will be needed soon at least for ARM64

Reply:
I added a paragraph on multi-arch to the doc.

@sosiouxme what additional information do you think we need to document here to get this to an approvable state?

Reply:
I think my concerns are answered.

ART can define a hook such that each time we build, we check if `installer/data/data/rhcos-stream.json` has changed. If it has, download the ISOs from the locations given there and make them available in the build context with an arch-specific name, probably something like `rhcos.$arch.iso` such that the same Dockerfile can just use the current arch to pick the right one in the arch-specific builds.